docs/documentation/features/frameworkUtilities.md

```diff
@@ -6,9 +6,116 @@ layout: documentation
 
 In this chapter useful services/utilities of the Eclipse SmartHome project are described. 
 
+## Network Server TLS service
+
+This service is responsible for providing SSLContexts that can be used all over ESH
+for server applications like webservlets, Mqtt servers or any extension that need to
+host a service. It can provide a SSLContext for TCP/TLS usage,
+as well as Connector for UDP/DTLS use cases.
+
+The SSLContext is configured to use either the user provided java keystore file
+or if none is provided to generate a self-signed certificate and public/private key-pair.
+
+The service can be asked to return different SSLContext or Connector objects based on a given
+context string. The default context will be returned if null is passed as context or if the
+given context is not configured.
+
+If nothing is configured (first-start), the service will generate a self-signed certificate and public/private key-pair in `ESH_DIR/etc/default.keystore` and store a new configuration file in `ESH_DIR/etc/networkServerTlsProvider.cfg`, including the keystore access password and private key password.
+
+You can provide your own java keystore files in `ESH_DIR/etc/yourKeystoreFile.keystore`. The name of the file is automatically the context name, e.g. you would request a `SSLContext` like this: `createSSLContext("yourKeystoreFile")`.
+Don't forget to add the configuration for your context to the configuration file as explained in the next section.
+
+The service can use ACME to automatically sign a public key by the Let's Encrypt initiative. 
+
+### Configuration
+
+The service accepts a configuration file `networkServerTlsProvider.cfg` with the following parameters:
+
+* **domains**: for instance "www.abc.org|default,www.def.org|second". *¹ *²
```

Comment:
@htreu To be honest I'm not happy with this solution as well. But let's face it, the OSGi properties based configuration is not useful for hierarchically organized configuration like for this or the Mqtt service. And the current ConfigAdmin implementation with non-removable (ghost) entries and no updates to the service on deleted files is not suitable for this or the Mqtt service. 

I would be very happy if a new configuration format (json based please, but I would accept ini file format as well) could be introduced. And if it is not centralized, that at least this (and maybe the Mqtt service) could read their own configuration files.

Reply:
Related to the discussion here: https://github.com/eclipse/smarthome/pull/3988. But I think for a security related service, there should be no ghost entries in this service due to ConfigAdmin if the file with the passwords (the cfg file) gets deleted.

Reply:
to be clear: do you plan to change the configuration structure once we implement #3988 like discussed? Deleting the cfg file is then the only thing left because ConfigAdmin will hold on to the current values. Do you think this is a restriction we can live with here?

Reply:
https://github.com/eclipse/smarthome/issues/4088#issuecomment-326270239
I will make the service a `ConfigurableService` implementation, for full PaperUI/GUI support. The current config structure would work. But as we agreed, it's not very user friendly.

`ConfigurableService` does not support tree structures or grouped, instance-able parameters. So I will need to discuss a change there.

Reply:
I have created an issue #4178

Reply:
The ConfigurableService does provide instance-able parameters: In PaperUI every service config does have the `Expert Mode` button where you can define arbitrary parameters for the service. Its not user friendly either but at least its a possibility to use the proposed structure (https://github.com/eclipse/smarthome/pull/4105/files#r136016624) just now. We can then think of an extended UI (discussion in #4178) later.

Reply:
I would say the solution is to remove the entire configuration parsing part here and use PR #4181 (Allow Configuration to handle nested structures and lists). That way I would just do:
`configuration = new Configuration(config).as(ServiceConfiguration.class);`

ConfigAdmin would need a change to provide me with a nested `config` map, according to #4178. Without this the service can be compiled and run but cannot be configured.

I also need support to write the marshalled config map back to the file. At the moment I'm doing this manually, but that causes other issues like that ConfigAdmin will send me the freshly synced configuration back and all the parsing happens in the background again.